pkg/querier/querier.go

```diff
@@ -1144,6 +1145,61 @@ func (q *SingleTenantQuerier) DetectedFields(ctx context.Context, req *logproto.
 	}, nil
 }
 
+func (q *SingleTenantQuerier) StructuredMetadata(ctx context.Context, req *logproto.StructuredMetadataKeysRequest) (*logproto.StructuredMetadataKeysResponse, error) {
+	expr, err := syntax.ParseLogSelector(req.Query, true)
+	if err != nil {
+		return nil, err
+	}
+	params := logql.SelectLogParams{
+		QueryRequest: &logproto.QueryRequest{
+			Start:     req.Start,
+			End:       req.End,
+			Limit:     1000,
+			Direction: logproto.BACKWARD,
+			Selector:  expr.String(),
+			Plan: &plan.QueryPlan{
+				AST: expr,
+			},
+		},
+	}
+
+	iters, err := q.SelectLogs(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(twhitney): we can improve this in the future as we don't need to iterate lines for structured metadata, just chunks
+	streams, err := streamsForFieldDetection(iters, 1000)
+	if err != nil {
+		return nil, err
+	}
+
+	structuredMetadata := parseStructuredMetadata(ctx, streams)
+
+	return &logproto.StructuredMetadataKeysResponse{
+		Keys: structuredMetadata,
+	}, nil
+}
+
+func parseStructuredMetadata(ctx context.Context, streams logqlmodel.Streams) []string {
+	keysm := map[string]struct{}{}
+
+	for _, stream := range streams {
+		for _, entry := range stream.Entries {
+			for _, l := range entry.StructuredMetadata {
+				keysm[l.Name] = struct{}{}
+			}
+		}
+	}
+
+	keys := make([]string, 0, len(keysm))
```

Comment:
you need to sort it I think to avoid unstable result ?

Reply:
It will get sorted here https://github.com/grafana/loki/pull/13119/files#diff-0670643837a81e9531f858f619539d92fc5be49b594b406d420215b308b61d97R1675-R1680 I think. Should be enough?